### PR TITLE
Stop skipping failing optimizer rules in tests

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/3695

 # Rationale for this change

Our tests currently ignore failing optimizer rules, so we are not catching regressions.

# What changes are included in this PR?

- Replace `SessionContext::new()` with a call to a new `create_test_ctx()` function, which does not skip failing optimizer rules
- Create a `create_test_ctx_skip_failing_optimizer_rules` function to use in the tests that would otherwise fail due to existing optimizer bugs:

```
    sql::group_by::csv_query_having_without_group_by
    sql::joins::equijoin_implicit_syntax_with_filter
    sql::joins::equijoin_left_and_condition_from_left
    sql::joins::equijoin_right_and_condition_from_left
    sql::joins::equijoin_right_and_condition_from_right
    sql::joins::reduce_left_join_1
    sql::joins::reduce_left_join_2
    sql::joins::reduce_left_join_3
    sql::joins::self_join_non_equijoin
    sql::select::csv_select_nested
    sql::select::csv_select_nested_without_aliases
    sql::select::unprojected_filter
    sql::timestamp::cast_string_to_time
```

# Are there any user-facing changes?

### CODE REVIEW

**Comment by andygrove** [comment]
@avantgardnerio you may be interested in this one

File: datafusion/core/tests/sql/mod.rs [hunk 1]
```diff
@@ -138,9 +139,22 @@ where
...
+    let config = SessionConfig::new().set_bool(OPT_OPTIMIZER_SKIP_FAILED_RULES, false);
+    SessionContext::with_config(config)
+}
+
+pub fn create_test_ctx_skip_failing_optimizer_rules() -> SessionContext {
+    // TODO we should not need this method but we need to fix the tests that depend on this first
+    // see https://github.com/apache/arrow-datafusion/issues/3695
+    let config = SessionConfig::new().set_bool(OPT_OPTIMIZER_SKIP_FAILED_RULES, true);
+    SessionContext::with_config(config)
+}
```
Comment (andygrove) [inline datafusion/core/tests/sql/mod.rs:153]: These are the new methods for creating the context

File: datafusion/core/tests/sql/select.rs [hunk 2]
```diff
@@ -996,7 +996,11 @@ async fn query_cte() -> Result<()> {
 
 #[tokio::test]
 async fn csv_select_nested() -> Result<()> {
-    let ctx = SessionContext::new();
+    // TODO we should not be ignoring optimizer errors here
+    // https://github.com/apache/arrow-datafusion/issues/3695
+    // Internal(\"Optimizer rule 'unwrap_cast_in_comparison' failed due to unexpected error:
+    //   Internal error: Error target data type UInt32
+    let ctx = create_test_ctx_skip_failing_optimizer_rules();
```
Comment (andygrove) [inline datafusion/core/tests/sql/select.rs:1003]: Here is an example where we do skip failing rules due to an existing bug

**Comment by avantgardnerio** [comment]
This could get tricky. The optimizers I wrote deliberately "failed" when they hit cases they didn't understand, because that was the agreed upon API at the time.

So they will definitely "fail" in edge cases, but that is by design. Maybe we need a specific error type for "decided not to optimize"?

**Comment by andygrove** [comment]
> Maybe we need a specific error type for "decided not to optimize"?

Perhaps have the `optimize()` method return a `Result<OptimizationOutcome>`, something like this?

```rust
enum OptimizationOutcome {
    /// The plan was successfully optimized
    Optimized(LogicalPlan),
    /// The rule was not capable of optimizing the plan 
    Skipped,
}
```

**Comment by alamb** [comment]
> Perhaps have the optimize() method return a Result<OptimizationOutcome>, something like this?

Sounds good to me -- maybe that would be the way to (eventually) be able to rerun multiple optimizer passes until they stopped being able to do anything

**Comment by alamb** [comment]
Do we still want to merge this PR?

**Comment by avantgardnerio** [comment]
> return a Result<OptimizationOutcome>

I'd prefer to use error types to determine control flow, because those allow us to use the `?` operator and keep syntax terse and readable. I think with all the recursive functions involved in optimizer rules, passing an enum up reliably will be burdensome and therefor either unreadable, or unlikely to be followed properly.

**Comment by andygrove** [comment]
Moving this to draft for now. I need to focus on fixing optimizer regressions for the 13.0.0 release but will come back to this issue once those are fixed. Thanks for the feedback so far @avantgardnerio and @alamb.

**Comment by xudong963** [comment]
> maybe that would be the way to (eventually) be able to rerun multiple optimizer passes until they stopped being able to do anything

+1, this is the ideal way.

**Pull request closed without merging**